receiver/telemetrygeneratorreceiver/otlp_generator.go

```diff
@@ -157,56 +156,139 @@ func (g *otlpGenerator) generateMetrics() pmetric.Metrics {
-						dps.At(l).SetTimestamp(pcommon.NewTimestampFromTime(getCurrentTime()))
+						updater(dps.At(l))
 					}
 				case pmetric.MetricTypeHistogram:
 					dps := metric.Histogram().DataPoints()
 					for l := 0; l < dps.Len(); l++ {
-						dps.At(l).SetTimestamp(pcommon.NewTimestampFromTime(getCurrentTime()))
+						updater(dps.At(l))
 					}
 				case pmetric.MetricTypeExponentialHistogram:
 					dps := metric.ExponentialHistogram().DataPoints()
 					for l := 0; l < dps.Len(); l++ {
-						dps.At(l).SetTimestamp(pcommon.NewTimestampFromTime(getCurrentTime()))
+						updater(dps.At(l))
 					}
 				}
 			}
 		}
 	}
+}
 
-	return g.metrics
+func (g *otlpGenerator) findLastMetricTime() time.Time {
+	maxTime := &time.Time{}
+	g.updateMetricTimes(func(ts timeStamped) {
+		if t := ts.Timestamp().AsTime(); t.After(*maxTime) {
+			*maxTime = t
+		}
+	})
+	return *maxTime
 }
 
-func (g *otlpGenerator) generateTraces() ptrace.Traces {
-	// calculate the time since the last baseline time we used to adjust the spans
+func (g *otlpGenerator) adjustMetricTimes() {
 	now := getCurrentTime()
-	timeSince := now.Sub(g.tracesStart)
-	// add the time since to the start and end times of all spans
+	maxTime := g.findLastMetricTime()
+	g.updateMetricTimes(func(ts timeStamped) {
+		delta := maxTime.Sub(ts.Timestamp().AsTime())
+		ts.SetTimestamp(pcommon.NewTimestampFromTime(now.Add(-delta)))
+	})
+	g.metricsStart = now
+}
+
+func (g *otlpGenerator) generateMetrics() pmetric.Metrics {
+	// calculate the time since the last baseline time we used to adjust the metrics
+	now := getCurrentTime()
+	timeSince := now.Sub(g.metricsStart)
+	g.updateMetricTimes(func(ts timeStamped) {
+		timeStamp := ts.Timestamp().AsTime().Add(timeSince)
+		ts.SetTimestamp(pcommon.NewTimestampFromTime(timeStamp))
+	})
+	// update the baseline time to the current time
+	g.metricsStart = now
+
+	return g.metrics
+}
+
+type timeStampedSpan interface {
+	StartTimestamp() pcommon.Timestamp
+	EndTimestamp() pcommon.Timestamp
+	SetStartTimestamp(pcommon.Timestamp)
+	SetEndTimestamp(pcommon.Timestamp)
+}
+
+type timeStampedSpanUpdater func(dp timeStampedSpan)
+
+func (g *otlpGenerator) updateTraceTimes(updater timeStampedSpanUpdater) {
 	for i := 0; i < g.traces.ResourceSpans().Len(); i++ {
 		resourceSpans := g.traces.ResourceSpans().At(i)
 		for k := 0; k < resourceSpans.ScopeSpans().Len(); k++ {
 			scopeSpans := resourceSpans.ScopeSpans().At(k)
 			for j := 0; j < scopeSpans.Spans().Len(); j++ {
 				span := scopeSpans.Spans().At(j)
-
-				startTime := span.StartTimestamp().AsTime().Add(timeSince)
-				span.SetStartTimestamp(pcommon.NewTimestampFromTime(startTime))
-
-				endTime := span.EndTimestamp().AsTime().Add(timeSince)
-				span.SetEndTimestamp(pcommon.NewTimestampFromTime(endTime))
+				updater(span)
 			}
 		}
 	}
+}
+
+// findLastTraceEndTime finds the span with the last end time
+func (g *otlpGenerator) findLastTraceEndTime() time.Time {
+	maxTime := &time.Time{}
```

Comment:
does this need to be a pointer? It looks like we dereference it everywhere it's used.

Reply:
You're right, the capture is already a reference